File: programs/r3-session-keys/tests/02-test_pause.rs
mod common;

use {
    common::{
        assert_program_paused, client, load, send_tx_expect_error, send_tx_expect_ok,
        timestamp_from_future, Env, DUMMY_ANCHOR_DISCRIMINATOR, TARGET_PROGRAM_PLACEHOLDER,
    },
    r3_session_keys::state::{
        program_config::{ProgramConfig, ProgramStatus},
        session::{Session, SessionStatus},
        user_smart_wallet::UserSmartWallet,
    },
    solana_keypair::Keypair,
    solana_signer::Signer,
};

#[test]
fn test_pause_blocks_creates_and_unpause_restores() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    let user_wallet = Keypair::new().pubkey();
    let other_user_wallet = Keypair::new().pubkey();
    let session_key = Keypair::new().pubkey();

    let (ix, user_smart_wallet, _) = client.create_smart_wallet(env.admin.pubkey(), user_wallet);
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin]);

    let ix = client.pause(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin]);
    assert_program_paused(&env);

    // program paused so other program functions should fail

    let (ix, _, _) = client.create_smart_wallet(env.admin.pubkey(), other_user_wallet);
    send_tx_expect_error(&mut env.svm, ix, &[&env.admin]);
    assert_program_paused(&env);

    let (ix, _, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key,
        TARGET_PROGRAM_PLACEHOLDER,
        timestamp_from_future(),
        DUMMY_ANCHOR_DISCRIMINATOR.to_vec(),
        8,
    );
    send_tx_expect_error(&mut env.svm, ix, &[&env.admin]);
    assert_program_paused(&env);

    let ix = client.unpause(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin]);
    let config: ProgramConfig = load(&env, &env.program_config);
    assert!(config.status == ProgramStatus::Active);

    // program unpaused so other program functions should succeed

    let (ix, other_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), other_user_wallet);
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin]);
    let _: UserSmartWallet = load(&env, &other_smart_wallet);

    let (ix, session, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key,
        TARGET_PROGRAM_PLACEHOLDER,
        timestamp_from_future(),
        DUMMY_ANCHOR_DISCRIMINATOR.to_vec(),
        8,
    );
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin]);
    let _: Session = load(&env, &session);
}

#[test]
fn test_revoke_session_works_while_paused() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    let user = Keypair::new();
    let session_key = Keypair::new().pubkey();

    let (ix, user_smart_wallet, _) = client.create_smart_wallet(env.admin.pubkey(), user.pubkey());
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin]);

    let (ix, session, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key,
        TARGET_PROGRAM_PLACEHOLDER,
        timestamp_from_future(),
        DUMMY_ANCHOR_DISCRIMINATOR.to_vec(),
        8,
    );
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin]);

    let ix = client.approve_session(user.pubkey(), user_smart_wallet, session_key);
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin, &user]);

    let ix = client.pause(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin]);
    assert_program_paused(&env);

    // Revocation is defensive and must not be blocked by the pause gate.
    let ix = client.revoke_session(user.pubkey(), user_smart_wallet, session_key);
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin, &user]);

    let state: Session = load(&env, &session);
    assert_eq!(state.status, SessionStatus::Revoked);
    assert_program_paused(&env);
}

File: tests/02-test_pause.ts
import * as anchor from "@anchor-lang/core";
import { Program } from "@anchor-lang/core";
import { Keypair, PublicKey, SystemProgram } from "@solana/web3.js";
import assert from "assert";

import { R3SessionKeys } from "../target/types/r3_session_keys";
import * as sessionKeys from "./utils/pda";
import {
  assertProgramActive,
  assertProgramPaused,
  ensureProgramInitialized,
  timestampFromFuture,
  sendExpectError,
} from "./utils/helpers";
import {
  DUMMY_ANCHOR_DISCRIMINATOR,
  ANCHOR_DISCRIMINATOR_LEN,
  TARGET_PROGRAM_PLACEHOLDER,
} from "./utils/constants";

describe("02 - Pause", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const baseWallet = provider.wallet as anchor.Wallet;

  const program = anchor.workspace.r3SessionKeys as Program<R3SessionKeys>;

  before(async () => {
    await ensureProgramInitialized(program, baseWallet);
  });

  function createSmartWallet(userWallet: PublicKey) {
    return program.methods
      .createSmartWallet(userWallet)
      .accounts({
        admin: baseWallet.publicKey,
      })
      .signers([baseWallet.payer]);
  }

  async function createSession(userSmartWallet: PublicKey, sessionKey: PublicKey) {
    return program.methods
      .createSession(
        sessionKey,
        TARGET_PROGRAM_PLACEHOLDER,
        await timestampFromFuture(provider.connection),
        DUMMY_ANCHOR_DISCRIMINATOR,
        ANCHOR_DISCRIMINATOR_LEN
      )
      .accounts({
        sessionExecutor: baseWallet.publicKey,
        userSmartWallet,
      })
      .signers([baseWallet.payer]);
  }

  it("Pause blocks creates and unpause restores", async () => {
    const userWallet = Keypair.generate().publicKey;
    const otherUserWallet = Keypair.generate().publicKey;
    const sessionKey = Keypair.generate().publicKey;

    const userSmartWalletPda = sessionKeys.deriveUserSmartWalletPda(program.programId, userWallet);
    await createSmartWallet(userWallet).rpc();

    await program.methods
      .pause()
      .accounts({
        admin: baseWallet.publicKey,
      })
      .signers([baseWallet.payer])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });
    await assertProgramPaused(program);

    // program paused so other program functions should fail

    await sendExpectError(createSmartWallet(otherUserWallet).rpc());
    await assertProgramPaused(program);

    await sendExpectError((await createSession(userSmartWalletPda, sessionKey)).rpc());
    await assertProgramPaused(program);

    await program.methods
      .unpause()
      .accounts({
        admin: baseWallet.publicKey,
      })
      .signers([baseWallet.payer])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });
    await assertProgramActive(program);

    // program unpaused so other program functions should succeed

    await createSmartWallet(otherUserWallet).rpc();
    const otherSmartWallet = await program.account.userSmartWallet.fetch(
      sessionKeys.deriveUserSmartWalletPda(program.programId, otherUserWallet)
    );
    assert.equal(otherSmartWallet.smartWalletOwner.toBase58(), otherUserWallet.toBase58());

    await (await createSession(userSmartWalletPda, sessionKey)).rpc();
    const session = await program.account.session.fetch(
      sessionKeys.deriveSessionPda(program.programId, userSmartWalletPda, sessionKey)
    );
    assert.equal(session.sessionKey.toBase58(), sessionKey.toBase58());
  });
});
